rpc/connection.go

```diff
@@ -271,54 +287,53 @@ func copyTLSConfig(c *tls.Config) *tls.Config {
-func NewTLSConnectionWithTLSConfig(srvAddr string, tlsConfig *tls.Config,
-	errorUnwrapper ErrorUnwrapper, handler ConnectionHandler,
-	connectNow bool, l LogFactory, wef WrapErrorFunc, log LogOutput,
-	tagsFunc LogTagsFromContext) *Connection {
+func NewTLSConnectionWithTLSConfig(
+	srvAddr string,
+	tlsConfig *tls.Config,
+	errorUnwrapper ErrorUnwrapper,
+	handler ConnectionHandler,
+	logFactory LogFactory,
+	logOutput LogOutput,
+	opts ConnectionOpts,
+) *Connection {
 	transport := &ConnectionTransportTLS{
 		srvAddr:    srvAddr,
 		tlsConfig:  copyTLSConfig(tlsConfig),
-		logFactory: l,
-		wef:        wef,
-	}
-	return NewConnectionWithTransport(handler, transport, errorUnwrapper,
-		connectNow, wef, log, tagsFunc)
-}
-
-// NewTLSConnectionWithProtocols returns a connection that tries to connect to
-// the given server address with TLS and registers custom protocols.
-func NewTLSConnectionWithProtocols(srvAddr string, rootCerts []byte,
-	errorUnwrapper ErrorUnwrapper, handler ConnectionHandler,
-	connectNow bool, l LogFactory, wef WrapErrorFunc, log LogOutput,
-	tagsFunc LogTagsFromContext, protocols []Protocol) *Connection {
-	transport := &ConnectionTransportTLS{
-		rootCerts:  rootCerts,
-		srvAddr:    srvAddr,
-		logFactory: l,
-		wef:        wef,
+		logFactory: logFactory,
+		wef:        opts.WrapErrorFunc,
 	}
-	return newConnectionWithTransportAndProtocols(handler, transport, errorUnwrapper,
-		connectNow, wef, log, tagsFunc, protocols)
+	return newConnectionWithTransportAndProtocols(handler, transport, errorUnwrapper, logOutput, opts)
 }
 
 // NewConnectionWithTransport allows for connections with a custom
 // transport.
-func NewConnectionWithTransport(handler ConnectionHandler,
-	transport ConnectionTransport, errorUnwrapper ErrorUnwrapper,
-	connectNow bool, wef WrapErrorFunc, log LogOutput,
-	tagsFunc LogTagsFromContext) *Connection {
-	return newConnectionWithTransportAndProtocols(handler, transport,
-		errorUnwrapper, connectNow, wef, log, tagsFunc, nil)
+func NewConnectionWithTransport(
+	handler ConnectionHandler,
+	transport ConnectionTransport,
+	errorUnwrapper ErrorUnwrapper,
+	logOutput LogOutput,
+	opts ConnectionOpts,
+) *Connection {
+	return newConnectionWithTransportAndProtocols(handler, transport, errorUnwrapper, logOutput, opts)
 }
 
 func newConnectionWithTransportAndProtocols(handler ConnectionHandler,
 	transport ConnectionTransport, errorUnwrapper ErrorUnwrapper,
-	connectNow bool, wef WrapErrorFunc, log LogOutput,
-	tagsFunc LogTagsFromContext, protocols []Protocol) *Connection {
-	// retry w/exponential backoff
-	reconnectBackoff := backoff.NewExponentialBackOff()
-	// never give up reconnecting
-	reconnectBackoff.MaxElapsedTime = 0
+	log LogOutput, opts ConnectionOpts) *Connection {
+	// use exponential backoffs by default which never give up on reconnecting
+	defaultBackoff := func() backoff.BackOff {
+		b := backoff.NewExponentialBackOff()
+		b.MaxElapsedTime = 0
```

Comment:
Not sure. The default is 15 seconds, so if we're ok with that we don't need to set anything.